packages/react-sdk/src/index.tsx

```diff
@@ -29,39 +66,93 @@ const Context = createContext<BucketContext>({
   flags: {
     flags: {},
     isLoading: false,
+    setContext: () => {},
+    context: {},
   },
 });
 
-export type BucketProps = BucketSDKOptions & {
-  publishableKey: string;
-  flags: FeatureFlagsOptions;
-  children?: ReactNode;
-  sdk?: BucketInstance;
-};
+export type BucketProps = BucketSDKOptions &
+  FlagContext & {
+    publishableKey: string;
+    flagOptions?: Omit<FeatureFlagsOptions, "context" | "fallbackFlags"> & {
+      fallbackFlags?: BucketFlags[];
+    };
+    children?: ReactNode;
+    sdk?: BucketInstance;
+  };
+
+export type Flags = { [k in BucketFlags]?: boolean };
 
-export default function Bucket({ children, sdk, ...config }: BucketProps) {
-  const [flags, setFlags] = useState<Flags>({});
+export function BucketProvider({
+  children,
+  sdk,
+  user: initialUser,
+  company: initialCompany,
+  otherContext: initialOtherContext,
+  ...config
+}: BucketProps) {
+  const [flags, setFlags] = useState<Flags>(
+    (config.flagOptions?.fallbackFlags ?? []).reduce((acc, key) => {
+      acc[key] = true;
+      return acc;
+    }, {} as Flags),
+  );
   const [flagsLoading, setFlagsLoading] = useState(true);
   const [bucket] = useState(() => sdk ?? BucketSingleton);
 
-  const contextKey = canonicalJSON(config);
+  const [flagContext, setFlagContext] = useState({
+    user: initialUser,
+    company: initialCompany,
+    otherContext: initialOtherContext,
+  });
+
+  const { user, company, otherContext } = flagContext;
+
+  // on mount
+  useEffect(() => () => bucket.init(config.publishableKey, config), []);
+  // on umount
+  useEffect(() => () => bucket.reset());
+
+  // if user.id or attributes change, send new attributes to the servers
   useEffect(() => {
-    try {
-      const { publishableKey, flags: flagOptions, ...options } = config;
+    if (user?.id) {
+      const { id, ...attributes } = user;
+      // `user` calls bucket.reset() automatically when needed
+      bucket.user(String(id), attributes);
+    }
+  }, [canonicalJSON({ user })]);
```

Comment:
Rather than relying on `useEffect`, it would probably be more elegant to create `useCallback` functions that you expose in the context. So that `useUser` calls a `setUser` from the context. `useEffect` is a bit of a crutch and it's easy to rely on it but in my experience most React errors are somehow related to `useEffect` triggering too frequently or not enough. https://react.dev/learn/you-might-not-need-an-effect

Reply:
changed them to useCallback as you suggested. Please have another look!